### PR TITLE
[dynamo] `ExecutorchCallDelegateHigherOrderVariable` - add sanity check that input and output tensors are disjoint

### LINKED ISSUE #111917: [dynamo] `ExecutorchCallDelegateHigherOrderVariable` may not preserve `FakeTensor` identity
### 🐛 Describe the bug

`ExecutorchCallDelegateHigherOrderVariable` will attempt to retrieve the real values as inputs to the executorch module, and then deep copy fake the result into new fake values. However, it is possible that the `executorch_call_delegate` function might perform in-place ops. 

The new FakeTensors would no longer share the same identity as the input fake tensors, violating our mechanism for Dynamo trace-time aliasing detection. 

No repro as of yet, this is just speculative

Solution: manually dedup by matching the `fake <-> real <-> fake` relationship between inputs and outputs.

### Versions

main

cc @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/111917

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/higher_order_ops.py [hunk 1]
```diff
@@ -694,7 +694,16 @@ def call_function(
...
+        real_args_tensor_ids = set(
+            pytree.tree_flatten_only(torch.Tensor, lambda t: id(t), real_sub_args)
+        )
+
         example_res = lowered_module.original_module(*real_sub_args)
+
+        real_res_tensor_ids = set(
+            pytree.tree_flatten_only(torch.Tensor, lambda t: id(t), example_res)
+        )
+        assert real_args_tensor_ids.isdisjoint(real_res_tensor_ids)
```
Comment (jon-chuang) [inline torch/_dynamo/variables/higher_order_ops.py:706]: Add comments/assert fail reason?

File: torch/_dynamo/variables/higher_order_ops.py [hunk 2]
```diff
@@ -81,6 +81,16 @@ def only_consist_of(var, types):
         return all(only_consist_of(item, types) for item in var.items.values())
     return False
 
+def _assert_tensors_nonaliasing(inputs, outputs):
```
Comment (jon-chuang) [inline torch/_dynamo/variables/higher_order_ops.py:84]: @tugsbayasgalan I've extracted into a helper function.

**Review by zou3519 — APPROVED** [review]

**Comment by izaitsevfb** [comment]
@pytorchbot revert -m "Fails internal executorch tests with module 'torch.utils._pytree' has no attribute 'tree_flatten_only'" -c ghfirst

``    _assert_tensors_nonaliasing(real_sub_args, example_res)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/buck-out/v2/gen/fbcode/9874f87d0ccb406e/executorch/backends/vulkan/test/__test_vulkan_delegate__/test_vulkan_delegate#link-tree/torch/_dynamo/variables/higher_order_ops.py", line 87, in _assert_tensors_nonaliasing
    pytree.tree_flatten_only(torch.Tensor, lambda t: id(t), inputs)
torch._dynamo.exc.InternalTorchDynamoError: module 'torch.utils._pytree' has no attribute 'tree_flatten_only'
```

**Comment by jon-chuang** [comment]
Oops, this code was not in pytorch testing path

**Comment by jon-chuang** [comment]
@angelayi @tugsbayasgalan maybe executorch is not functional as claimed? 

I can't inspect the internal tests

**Comment by izaitsevfb** [comment]
> @angelayi @tugsbayasgalan maybe executorch is not functional as claimed?
> 
> I can't inspect the internal tests

The tests fail the added check:
`AssertionError: inputs to function body cannot alias outputs`. I've created an internal working group with relevant PoCs to discuss the failure.

**Comment by jon-chuang** [comment]
Cool. Either
1. Executorch is made functional
2. We will try to dedup the fake tensors of input-output aliases

**Comment by izaitsevfb** [comment]
@angelayi volunteered to debug the test failure (huge thanks!). Meanwhile I'd have to temporarily revert this PR and #111938 on top of it. Apologies for the churn.

**Comment by jon-chuang** [comment]
> maybe executorch is not functional as claimed?

@angelayi any findings on this so far? 

Would like to get this check in place eventually, it's not high prio but will help close the book on the FakeTensor object identity efforts.

**Pull request closed without merging**